src/Core/Usecase/Message/ReceiveMessage.php

```diff
@@ -145,89 +125,131 @@ private function createIncomingMessage($incoming_message, $contact_id, $survey_s
-		$messageInSurveyState = $messageInSurveyState->get($surveyStateEntity->message_id);
+		// Load the survey state for this contact
+		$surveyStateEntity = $this->targeted_survey_state_repo->getActiveByContactId($incoming_message->contact_id);
+
+		// Attempt to load the previous message
+		$messageInSurveyState = $this->repo->get($surveyStateEntity->message_id);
+
+		// If we didn't find an outgoing message
 		if (!$messageInSurveyState || $messageInSurveyState->direction !== \Ushahidi\Core\Entity\Message::OUTGOING) {
-			//we can't save it as a message of the survey
-			Kohana::$log->add(
-				Log::ERROR,
-				'Could not add contact\'s  message for contact_id: ' .
-				print_r($contact_id, true) . ' and form ' . $surveyStateEntity->form_id
-			);
-			throw new HTTP_Exception_400(
-				'Outgoing question not found for contact ' . $contact_id . ' and form ' . $surveyStateEntity->form_id
+			// We can't save it as a message of the survey
+			// ... log an error because we should probably never end up here
+			\Kohana::$log->add(
+				\Log::ERROR,
+				'Could not add contact\'s  message',
+				[
+					'contact_id' => $incoming_message->contact_id,
+					'form_id' => $surveyStateEntity->form_id
+				]
 			);
+
+			// Create a new post
+			$post_id = $this->createPost($incoming_message);
+			$incoming_message->setState(compact('post_id'));
+
+			// But always save the message anyway - otherwise its lost forever
+			return $this->repo->create($incoming_message);
 		}
-		//get the next attribute in that form, based on the form and the last_sent_form_attribute_id
+
+		// We found the outgoing message... flow continues
+		// Save the incoming message
+		$incomingMessageId = $this->createIncomingMessage($incoming_message, $surveyStateEntity);
+
+		// Get the next attribute in that form, based on the form and the last_sent_form_attribute_id
 		$next_form_attribute = $this->form_attr_repo->getNextByFormAttribute(
 			$surveyStateEntity->form_attribute_id
 		);
-		//create incoming message
-		$incomingMessageId = $this->createIncomingMessage($incoming_message, $contact_id, $surveyStateEntity);
-		// intermediate state to mark when we receive a message
+
+		// Set up intermediate state w/ message id, next attribute and new status
 		$surveyStateEntity->setState(
 			[
 				'form_attribute_id' => $next_form_attribute->getId(),
 				'message_id' => $incomingMessageId,
 				'survey_status' => Entity\TargetedSurveyState::RECEIVED_RESPONSE
 			]
 		);
+		// And save intermediate state
 		$this->targeted_survey_state_repo->update($surveyStateEntity);
+
+		// If we have another question to send
 		if ($next_form_attribute->getId() > 0) {
-			$newMessageId = $this->createOutgoingMessage($contact_id, $surveyStateEntity, $next_form_attribute);
+			// Queue next question to be sent
+			$outgoingMessageId = $this->createOutgoingMessage(
+				$incoming_message->contact_id,
+				$surveyStateEntity,
+				$next_form_attribute,
+				$incoming_message->data_provider
```

Comment:
We're now using the incoming data provider when sending the outgoing. ie. if we receive a message from Twilio, always reply with Twilio

Reply:
Yeah, I like that better. I was thinking of doing the same thing, but was also debating about specifying SMS provider for each survey, as described in my first comment [here](https://github.com/ushahidi/platform/issues/2842), though it potentially breaks a lot of things and requires a migration to `forms`.